rfcs/xxxx-collage.md

```diff
@@ -0,0 +1,987 @@
+
+The user can use `on_device` annotations to constrain sub-graphs to particular devices. When Collage is considering
+candidate partitions, it should be sure to choose a candidate `Target` which 'refines' the `Target` for every
+sub-expression discovered by the `PlanDevicesPass`. Given targets T and U we say 'T refines U' if T has a
+'"compiler"' and/or '"partition_rule"' attributes, U has no such attributes, and T and U otherwise agree on all other
+fields.
+
+### Phase 2
+
+Most of the hard work for this phase is carried by the `AllCandidates` implementations of the `PartitionRule`s. The main
+driver simply needs to index all the found `CandidatePartitions` by their minimum 'inside' `PostDfsIndex`
+for rapid retrieval during the shortest path search.
+
+### Phase 3
+
+We find it most natural to use Dijkstra to find the optimal partitioning. A `SearchState` is:
+
+- An `IndexSet` of the dataflow nodes already 'covered' by candidates on the best path to this state. This is the
+  identifying key for the state.
+- The predecessor `SearchState` in the best path to this state.
+- The `Cost` of the best path to this state. This is the order for the Dijkstra priority queue.
+- The `CandidatePartition` for the transition from the best predecessor to this state.
+
+The starting state has no covered nodes. The final state has all nodes covered.
+
+When expanding a state we could choose any `CandidatePartition` collected from phase 2 provided it doesn't overlap with
+the state's covered set. However, a search path applying candidates C then D is equivalent to one applying D then C, so
+we only consider candidates which intersect the next yet-to-be-covered dataflow node. For each such candidate we use
+the `CostEstimator` (with it's assumed cache) to get the candidate's cost, build the successor state, and 'relax' the
+successor state in the usual way. (See Appendix E for more details on `CostEstimator`.)
+
+The `HostPartitionRule` is used to allow some dataflow nodes to be 'left behind' for execution by the host.
+
+### Phase 4
+
+The overall Relay expression is partitioned over all the `CandidatePartition`s on the shortest path 'in parallel'. Since
+all the candidates are expressed using `SubGraph`s w.r.t. the original dataflow graph, we must be careful not to
+invalidate yet-to-be-partitioned candidates as we go. Working backwards in dataflow order avoids this problem.
+
+## Known Limitations
+
+- **Some BYOC boilerplate changes required**: TVM's current BYOC integration API only requires the 'lowering/codegen'
+  function to be registered to a well-known global function name. Everything else is up to the BYOC author.
+    - Collage requires pattern-based BYOC integrations to register their patterns in the global pattern table. Some BYOC
+      integrations use the table but many do not, but it's an easy fix.
+    - Collage requires the BYOC lowering function to yield a valid `runtime::Module` without requiring any additional
```

Comment:
Will Collage work with Target Hooks ?

Reply:
It will need to, yes, I've just not gotten around to figuring out the encoding so that the hooks make the correct redirection, and how to tell  the difference between using Compiler=<toolchain> vs Target=.... But I think it's just leg work.

Reply:
Added a TODO for this.